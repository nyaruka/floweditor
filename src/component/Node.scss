@import "../variables.scss";

.node_container {
    padding: $endpoint_buffer;
    z-index: $z_nodes;
    position: absolute;

    .node {
        border: 1px solid $borders;
        border-radius: $curvature;
        background: #fff;
        padding: 0;
        -webkit-user-select: none;
        -moz-user-select: none;
        -ms-user-select: none;
        user-select: none;
        transition: all 100ms ease-in-out;
        cursor: pointer;
        display: table;
        
        &.dragging {
            box-shadow: $node_shadow $node_shadow_color;
            z-index: $z_dragging;
            .add {
                visibility: hidden !important;
            }
            transition: none;
            border-radius: $curvature;
        }

        &.ghost {
            opacity: .5;
            transition: none;
        }

        &.nondragged {
            opacity: .6;
            transition: opacity 400ms ease;
        }

        .cropped {
            box-shadow: -1px -1px 0px 0px $node_shadow_color;
            overflow: hidden;
            min-width: $node_width;
            border-top-left-radius: $curvature;
            border-top-right-radius: $curvature;
        }

        .actions {
            width: $node_width;
            background: #fff;
        }

        &.translating {

            .actions {
                border-bottom-left-radius: 7px;
                border-bottom-right-radius: 7px;
                overflow: hidden;
            }

            .unnamed_exit {
                &.exit_table {
                    height: inherit;
                    border-top: none;
                    
                    .exits {
                        height: inherit;
                    }
                    .exit {
                        padding: 0 !important;
                        margin: 0;
                    }
                }
            }
        }

        .exit_table {
            display: table;
            width: 100%;
            box-shadow: -1px -1px 0px 0px $node_shadow_color;   
            border-bottom-left-radius: $curvature;
            border-bottom-right-radius: $curvature;

            .exits {
                border-collapse: collapse;
                display: table-row;
                height: 7px;
                &:after {
                    content: "";
                    display: table;
                    clear: both;
                }
            }
        }
        .drag_group {
            position: absolute;
            visibility: hidden;
            bottom: $endpoint_buffer;
            left: $endpoint_buffer;
            padding: 3px;
            color: $light_gray - #222;
            font-size: 8px;
            text-decoration: none;
            cursor: move;
            z-index: $z_group_drag_anchor;
            &:hover {
                color: $light-gray - #333;
                border-color: $light_gray - #222;
            }
        }
        .add {
            position: absolute;
            right: $endpoint_buffer / 2;
            bottom: $endpoint_buffer / 2;
            visibility: hidden;
            color: $blue;
            text-decoration: none;
            background: #fff;
            -webkit-border-radius: 5px;
            -moz-border-radius: 5px;
            border-radius: 5px;
            line-height: 8px;
            font-size: 14px;
        }
        &:hover {
            .add {
                visibility: visible;
            }
            .drag_group {
                visibility: visible;
            }
        }
    }
}

.active {
    position: absolute;
    background: $active_color;
    border: 1px solid $active_color - #111;
    -webkit-border-radius: 12px;
    -moz-border-radius: 12px;
    border-radius: 12px;
    padding: 3px 5px;
    color: #fff;
    font-weight: 500;
    top: -10px;
    left: -10px;
    font-size: 13px;
    min-width: 22px;
    text-align: center;
    cursor: default;
}
