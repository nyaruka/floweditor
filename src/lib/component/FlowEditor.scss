@import "../../variables.scss";

.editor {
    position: relative;
    font-family: 'Roboto', sans-serif;
    font-weight: 300;
    color: $text_color; // height: 100vh;
    height: 20000px;
    font-size: 13px;
    line-height: 1.2;
    margin: 0;
    background-color: rgba(80, 80, 80, .02);
    background-position: -10px -10px;
    $color_1: rgba(80, 80, 80, 0.05);
    background-image: linear-gradient(0deg, transparent 24%, $color_1 25%, $color_1 26%, transparent 27%, transparent 74%, $color_1 75%, $color_1 76%, transparent 77%, transparent), linear-gradient(90deg, transparent 24%, $color_1 25%, $color_1 26%, transparent 27%, transparent 74%, $color_1 75%, $color_1 76%, transparent 77%, transparent);
    background-size: 40px 40px;
}

.translating {
    .editor {
        $color_1: rgba(0, 0, 0, .08);
        $color_2: rgba(0, 0, 0, .05);
        background: repeating-linear-gradient(-60deg, $color_1, $color_1 3px, $color_2 3px, $color_2 6px); // background: rgba(200, 200, 200, .3);
    }
    svg path {
        stroke: $red !important;
        fill: $red !important;
    }
}
