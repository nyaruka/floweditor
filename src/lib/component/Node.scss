@import "../../variables.scss";

.node {
    padding: 0;
    &.dragging {
        z-index: $z_dragging; // box-shadow: 0 0 0 4px rgba(0, 0, 0, .1), 0 0 0 1px rgba(0, 0, 0, .1) inset; // transform: scale(1.05); // border: 1px solid #ddd;
        .add {
            visibility: hidden !important;
        }
        transition: none; // transform: scale(1.05);
        box-shadow: 5px 5px 0px -2px rgba(0, 0, 0, .06);
        transform: scale(1.015);
    }
    -webkit-user-select: none;
    -moz-user-select: none;
    -ms-user-select: none;
    user-select: none;
    transition: top 200ms ease;
    position: absolute; // box-shadow: 0 0 0 1px rgba(0, 0, 0, .1) inset; // border: 1px solid ;
    width: $node_width;
    z-index: $z_nodes;
    cursor: pointer; // box-shadow: 0px 0px 1px 0px rgba(0, 0, 0, .4);
    display: table;
    &.ghost {
        opacity: .5;
        transition: none;
    }
    .actions {
        border-radius: 25px;
        width: 200px;
        &.exit_table {
            border-top: none;
            .exit {
                padding: 0 !important;
                margin: 0;
            }
        }
        .action:last-child {
            padding-bottom: 15px;
        }
    }
    .exit_table {
        background: #fff;
        display: table;
        width: 198px; //border-top: 1px solid #f6f6f6;
        border-left: 1px solid $borders;
        border-right: 1px solid $borders;
        border-bottom: 1px solid $borders;
        .exits {
            border-collapse: collapse;
            display: table-row;
            &:after {
                content: "";
                display: table;
                clear: both;
            }
        }
    }
    .drag_group {
        position: absolute;
        visibility: hidden;
        bottom: 0px;
        left: 0px;
        padding: 3px;
        color: $light_grey - #222;
        font-size: 8px;
        text-decoration: none;
        cursor: move;
        z-index: $z_group_drag_anchor;
        &:hover {
            color: $light-grey - #333;
            border-color: $light_grey - #222;
        }
    }
    .add {
        position: absolute;
        right: -5px;
        bottom: -5px;
        visibility: hidden;
        color: $blue;
        text-decoration: none;
        background: #fff;
        -webkit-border-radius: 5px;
        -moz-border-radius: 5px;
        border-radius: 5px;
        line-height: 8px;
        font-size: 14px;
    }
    &:hover {
        .add {
            visibility: visible;
        }
        .drag_group {
            visibility: visible;
        }
    }
}

.active {
    position: absolute;
    background: $active_color;
    border: 1px solid $active_color - #111;
    -webkit-border-radius: 12px;
    -moz-border-radius: 12px;
    border-radius: 12px;
    padding: 3px 5px;
    color: #fff;
    font-weight: 500;
    top: -10px;
    left: -10px;
    font-size: 13px;
    min-width: 22px;
    text-align: center;
    cursor: default;
}
