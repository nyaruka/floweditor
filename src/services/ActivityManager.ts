import { Exit } from '../flowTypes';
import { GetActivity } from '../providers/ConfigProvider/external';
import Counter from '../components/Counter';

// how often we ask the server for new data
const REFRESH_SECONDS = 10;

/**
 * Contains all the activity data for a flow
 */
export interface Activity {
    // exit_uuid:destination_node_uuid -> count
    segments: { [key: string]: number };

    // node_uuid -> count
    nodes: { [key: string]: number };
}

export default class ActivityManager {
    private static singleton: ActivityManager;

    // our main activity fetch from the external
    private activity: Activity;

    // our simulation activity
    private simulation: Activity;

    private flowUUID: string;
    private getActivityExternal: GetActivity;

    private listeners: { [key: string]: Counter } = {};
    private timer: any;

    constructor(flowUUID: string, getActivity: GetActivity) {
        this.flowUUID = flowUUID;
        this.getActivityExternal = getActivity;

        this.clearSimulation = this.clearSimulation.bind(this);
        this.setSimulation = this.setSimulation.bind(this);
        this.fetchActivity = this.fetchActivity.bind(this);
        this.notifyListeners = this.notifyListeners.bind(this);
        this.deregister = this.deregister.bind(this);
        this.registerListener = this.registerListener.bind(this);
        this.getActivity = this.getActivity.bind(this);
        this.getActiveCount = this.getActiveCount.bind(this);
        this.getPathCount = this.getPathCount.bind(this);

        this.fetchActivity();
    }

    public clearSimulation() {
        this.simulation = null;
        this.notifyListeners();
        this.fetchActivity();
    }

    public setSimulation(activity: Activity) {
        this.simulation = activity;
        if (this.timer) {
            window.clearTimeout(this.timer);
            this.timer = null;
        }
        this.notifyListeners();
    }

    private fetchActivity(wait: number = 0) {
        // if (!this.timer) {
        //     this.timer = window.setTimeout(() => {
        //         this.timer = null;
        //         this.getActivityExternal(this.flowUUID)
        //             .then((activity: Activity) => {
        //                 this.activity = activity;
        //                 this.notifyListeners();
        //             })
        //             .catch(() => {
        //                 // ignore missing activity
        //             });

        //         this.fetchActivity(REFRESH_SECONDS * 1000);
        //     }, wait);
        // }
    }

    public notifyListeners() {
        for (let counter in this.listeners) {
            this.listeners[counter].requestUpdate();
        }
    }

    public deregister(key: string) {
        delete this.listeners[key];
    }

    public registerListener(counter: Counter) {
        // this is called from ref, which is null on unmounts
        if (counter) {
            this.listeners[counter.getKey()] = counter;
        }
    }

    public getActivity(): Activity {
        if (this.simulation) {
            return this.simulation;
        }
        return this.activity;
    }

    public getActiveCount(nodeUUID: string): number {
        var activity = this.getActivity();
        if (activity) {
            var count = activity.nodes[nodeUUID];
            if (count !== undefined) {
                return count;
            }
        }
        return 0;
    }

    public getPathCount(exit: Exit): number {
        var activity = this.getActivity();
        if (activity) {
            if (exit.destination_node_uuid) {
                var count = activity.segments[exit.uuid + ':' + exit.destination_node_uuid];
                if (count !== undefined) {
                    return count;
                }
            }
        }
        return 0;
    }
}
