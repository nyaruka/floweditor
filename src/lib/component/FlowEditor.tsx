import * as React from 'react';
import { react as bindCallbacks } from 'auto-bind';
import { jsonEqual } from '../utils';
import { getFlow, getFlows, FlowDetails } from '../external';
import { FlowDefinition, Languages } from '../flowTypes';
import FlowMutator from '../services/FlowMutator';
import Temba from '../services/Temba';
import ComponentMap from '../services/ComponentMap';
import Flow from './Flow';
import FlowList, { FlowOption } from './FlowList';
import LanguageSelectorComp, { Language } from './LanguageSelector';
import { languagesPT, endpointsPT, ConfigProvider } from '../config';

import * as styles from './FlowEditor.scss';

export type OnSelectFlow = ({ uuid }: FlowOption) => void;

export interface FlowEditorProps {
    config: FlowEditorConfig;
}

export interface EditorState {
    fetching: boolean;
    language: Language;
    baseLanguage: Language;
    translating: boolean;
    nodeDragging: boolean;
    definition: FlowDefinition;
    flows: Array<{ uuid: string; name: string }>;
    dependencies: FlowDefinition[];
}

export const getBaseLanguage = (languages: {
    [iso: string]: string;
}): Language => {
    const [iso] = Object.keys(languages);
    const name = languages[iso];
    return {
        name,
        iso
    };
};

/**
 * A navigable list of flows for an account
 */
export default class Editor extends React.PureComponent<
    FlowEditorProps,
    EditorState
> {
    private Temba: Temba;
    private Mutator: FlowMutator;
    private ComponentMap: ComponentMap;

    public static contextTypes = {
        languages: languagesPT,
        endpoints: endpointsPT
    };

    constructor(props: FlowEditorProps) {
        super(props);

        const language = getBaseLanguage(props.config.languages);

        this.state = {
            language,
            baseLanguage: language,
            translating: false,
            fetching: false,
            nodeDragging: false,
            definition: null,
            flows: [],
            dependencies: null
        };

        // this.Temba = new Temba('https://your-site.com', '05594lM5uQsTHLlvrBts5lenb5Iyex6P');

        bindCallbacks(this, {
            include: [
                'onSelectFlow',
                'setLanguage',
                'setDefinition',
                'onDrag',
                'save'
            ]
        });
    }

    public componentDidMount(): void {
        this.fetchFlow(this.props.config.flow);
        this.fetchFlowList();
    }

    private onSelectFlow({ uuid }: { uuid: string; name: string }): void {
        if (this.state.flows.length) {
            if (uuid !== this.state.definition.uuid) {
                this.setState({ fetching: true }, () => this.fetchFlow(uuid));
            }
        }
    }

    private onDrag(dragging: boolean): void {
        if (this.state.nodeDragging !== dragging) {
            this.setState({ nodeDragging: dragging });
        }
    }

    private setDefinition(
        definition: FlowDefinition,
        dependencies?: FlowDefinition[]
    ): void {
        // console.log(`\nDEFINITION: \n${JSON.stringify(definition, null, 2)}\n`);

        if (dependencies) {
            this.setState({
                fetching: false,
                definition,
                dependencies
            });
        } else {
            this.setState({
                fetching: false,
                definition
            });
        }
    }

    private save(definition: FlowDefinition): void {
        // this.props.External.saveFlow(definition).catch((error) => {
        // do nothing
        // });
    }

    // TODO: determine full dependency list and fetch those at simulation time
    private initialize(definition: FlowDefinition): void {
        this.ComponentMap = new ComponentMap(definition);

        this.Mutator = new FlowMutator(
            this.ComponentMap,
            definition,
            this.setDefinition,
            this.save
            // this.props,
        );

        this.setDefinition(definition);
    }

    private fetchFlow(uuid: string): void {
        getFlow(this.props.config.endpoints.flows, uuid, false)
            .then(({ definition }: FlowDetails) => this.initialize(definition))
            .catch((error: any) => console.log(`fetchFlow error: ${error}`));
    }

    private fetchFlowList(): void {
        getFlows(this.props.config.endpoints.flows)
            .then((flows: FlowDetails[]) =>
                this.setState({
                    flows: flows.map(({ uuid, name }) => ({
                        uuid,
                        name
                    }))
                })
            )
            .catch((error: any) =>
                console.log(`fetchFlowList error: ${error}`)
            );
    }

    private setLanguage(language: Language): void {
        const translating: boolean = !jsonEqual(
            this.state.baseLanguage,
            language
        );

        this.setState({
            language,
            translating
        });
    }

    private getLanguageSelector(): JSX.Element {
        if (this.props.config.languages) {
            return (
                <LanguageSelectorComp
                    languages={this.props.config.languages}
                    iso={this.state.language.iso}
                    onChange={this.setLanguage}
                />
            );
        }
        return null;
    }

    private getFlow(): JSX.Element {
        const displayFlow = this.state.definition && !this.state.fetching;
        if (displayFlow) {
            return (
                <Flow
                    nodeDragging={this.state.nodeDragging}
                    onDrag={this.onDrag}
                    language={this.state.language}
                    translating={this.state.translating}
                    definition={this.state.definition}
                    dependencies={this.state.dependencies}
                    ComponentMap={this.ComponentMap}
                    Mutator={this.Mutator}
                />
            );
        }
        return null;
    }

    public render(): JSX.Element {
        const languageSelector: JSX.Element = this.getLanguageSelector();

        const flowOption: FlowOption = this.state.definition
            ? {
                  uuid: this.state.definition.uuid,
                  name: this.state.definition.name
              }
            : null;

        const flow: JSX.Element = this.getFlow();

        const translatingStyle = this.state.translating
            ? styles.translating
            : null;

        return (
            <ConfigProvider
                languages={this.props.config.languages}
                endpoints={this.props.config.endpoints}>
                <div className={translatingStyle} data-spec="editor-container">
                    <div className={styles.editor} data-spec="editor">
                        <FlowList
                            flowOption={flowOption}
                            flowOptions={this.state.flows}
                            onSelectFlow={this.onSelectFlow}
                        />
                        {languageSelector}
                        {flow}
                    </div>
                </div>
            </ConfigProvider>
        );
    }
}
