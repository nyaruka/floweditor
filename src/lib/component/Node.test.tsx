import * as React from 'react';
import { shallow } from 'enzyme';
import CompMap from '../services/ComponentMap';
import Plumber from '../services/Plumber';
import ActivityManager from '../services/ActivityManager';
import Node, { NodeProps, getLocalizations } from './Node';
import { getBaseLanguage } from './FlowEditor';

const config = require('../../../assets/config.json');
const {
    results: [{ uuid: flowUUID, definition }]
} = require('../../test_flows/a4f64f1b-85bc-477e-b706-de313a022979.json');

const language = getBaseLanguage(config.languages);

const {
    nodes: [, node],
    localization: locals,
    _ui: { nodes: uiNodes }
} = definition;
const Activity = new ActivityManager(flowUUID, jest.fn());
const ComponentMap = new CompMap(definition);
const props: NodeProps = {
    ComponentMap,
    language,
    translating: false,
    definition,
    onNodeMounted: jest.fn(),
    onUpdateDimensions: jest.fn(),
    onNodeMoved: jest.fn(),
    onNodeDragStart: jest.fn(),
    onNodeBeforeDrag: jest.fn(),
    onDisconnectExit: jest.fn(),
    onNodeDragStop: jest.fn(),
    onAddAction: jest.fn(),
    onRemoveAction: jest.fn(),
    onRemoveNode: jest.fn(),
    node,
    onUpdateAction: jest.fn(),
    onUpdateRouter: jest.fn(),
    onUpdateLocalizations: jest.fn(),
    openEditor: jest.fn(),
    onMoveActionUp: jest.fn(),
    ui: uiNodes[node.uuid],
    Activity,
    translations: null,
    plumberDraggable: jest.fn(),
    plumberMakeTarget: jest.fn(),
    plumberRemove: jest.fn(),
    plumberRecalculate: jest.fn(),
    plumberMakeSource: jest.fn(),
    plumberConnectExit: jest.fn()
};
const iso = 'spa';
const translations = locals[iso];

const { languages } = config;

describe('Node', () => {
    describe('getLocalizations', () => {
        it('should compose a list of localizations', () =>
            expect(
                getLocalizations(node, iso, languages, translations)
            ).toMatchSnapshot());
    });

    describe('Component: NodeComp', () => {
        it('should render', () => {
            const wrapper = shallow(<Node {...props} />, {
                context: { languages }
            });

            expect(wrapper.exists()).toBeTruthy();
            expect(wrapper.state('dragging')).toBeFalsy();
        });
    });
});
