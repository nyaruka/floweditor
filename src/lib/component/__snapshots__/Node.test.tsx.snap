// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Node getLocalizations should compose a list of localizations 1`] = `
Array [
  LocalizedObject {
    "iso": "spa",
    "language": Object {
      "iso": "spa",
      "name": "Spanish",
    },
    "languages": Object {
      "eng": "English",
      "fre": "French",
      "spa": "Spanish",
    },
    "localized": true,
    "localizedKeys": Object {
      "arguments": true,
    },
    "localizedObject": Object {
      "arguments": Array [
        "rojo, r",
      ],
      "exit_uuid": "55855afc-f612-4ef9-9288-dcb1dd136052",
      "type": "has_any_word",
      "uuid": "fa0a9b24-5f19-4b8e-b287-27af5811de1d",
    },
  },
  LocalizedObject {
    "iso": "spa",
    "language": Object {
      "iso": "spa",
      "name": "Spanish",
    },
    "languages": Object {
      "eng": "English",
      "fre": "French",
      "spa": "Spanish",
    },
    "localized": true,
    "localizedKeys": Object {
      "arguments": true,
    },
    "localizedObject": Object {
      "arguments": Array [
        "naranja, n",
      ],
      "exit_uuid": "668ca2ab-8d49-47f5-82a1-e3a82a58e5fb",
      "type": "has_any_word",
      "uuid": "b5f900b9-ad13-479a-8ad3-1f1ad5ac88f2",
    },
  },
  LocalizedObject {
    "iso": "spa",
    "language": Object {
      "iso": "spa",
      "name": "Spanish",
    },
    "languages": Object {
      "eng": "English",
      "fre": "French",
      "spa": "Spanish",
    },
    "localized": true,
    "localizedKeys": Object {
      "arguments": true,
    },
    "localizedObject": Object {
      "arguments": Array [
        "amarillo, a",
      ],
      "exit_uuid": "14806949-d583-49e2-aa55-03aa16ee5a3a",
      "type": "has_any_word",
      "uuid": "e9c842e8-f1c5-4f07-97e7-50a4f93b22e5",
    },
  },
  LocalizedObject {
    "iso": "spa",
    "language": Object {
      "iso": "spa",
      "name": "Spanish",
    },
    "languages": Object {
      "eng": "English",
      "fre": "French",
      "spa": "Spanish",
    },
    "localized": false,
    "localizedKeys": Object {},
    "localizedObject": Object {
      "arguments": Array [
        "green, g",
      ],
      "exit_uuid": "77394377-f6b8-4366-9bef-d468621258ef",
      "type": "has_any_word",
      "uuid": "cc5894af-5dce-454e-a525-3d7c5c41d21d",
    },
  },
  LocalizedObject {
    "iso": "spa",
    "language": Object {
      "iso": "spa",
      "name": "Spanish",
    },
    "languages": Object {
      "eng": "English",
      "fre": "French",
      "spa": "Spanish",
    },
    "localized": false,
    "localizedKeys": Object {},
    "localizedObject": Object {
      "arguments": Array [
        "blue, b",
      ],
      "exit_uuid": "92d429d8-c275-4306-9360-93f4b9c7acb1",
      "type": "has_any_word",
      "uuid": "590d13e3-7b47-44e3-b8a0-ba9bd41d75d2",
    },
  },
  LocalizedObject {
    "iso": "spa",
    "language": Object {
      "iso": "spa",
      "name": "Spanish",
    },
    "languages": Object {
      "eng": "English",
      "fre": "French",
      "spa": "Spanish",
    },
    "localized": false,
    "localizedKeys": Object {},
    "localizedObject": Object {
      "arguments": Array [
        "indigo, i",
      ],
      "exit_uuid": "2de9af80-1bd9-4f37-839f-073edbd14369",
      "type": "has_any_word",
      "uuid": "2a7cbed1-6597-4545-b145-14a2e9282e6c",
    },
  },
  LocalizedObject {
    "iso": "spa",
    "language": Object {
      "iso": "spa",
      "name": "Spanish",
    },
    "languages": Object {
      "eng": "English",
      "fre": "French",
      "spa": "Spanish",
    },
    "localized": false,
    "localizedKeys": Object {},
    "localizedObject": Object {
      "arguments": Array [
        "violet, v",
      ],
      "exit_uuid": "5760ec2f-04d4-492b-817b-9f395633ec79",
      "type": "has_any_word",
      "uuid": "ab99e18c-433f-436e-9278-08bcf506f433",
    },
  },
  LocalizedObject {
    "iso": "spa",
    "language": Object {
      "iso": "spa",
      "name": "Spanish",
    },
    "languages": Object {
      "eng": "English",
      "fre": "French",
      "spa": "Spanish",
    },
    "localized": true,
    "localizedKeys": Object {
      "name": true,
    },
    "localizedObject": Object {
      "destination_node_uuid": "bc978e00-2f3d-41f2-87c1-26b3f14e5925",
      "name": Array [
        "rojo",
      ],
      "uuid": "55855afc-f612-4ef9-9288-dcb1dd136052",
    },
  },
  LocalizedObject {
    "iso": "spa",
    "language": Object {
      "iso": "spa",
      "name": "Spanish",
    },
    "languages": Object {
      "eng": "English",
      "fre": "French",
      "spa": "Spanish",
    },
    "localized": true,
    "localizedKeys": Object {
      "name": true,
    },
    "localizedObject": Object {
      "destination_node_uuid": "bc978e00-2f3d-41f2-87c1-26b3f14e5925",
      "name": Array [
        "naranja",
      ],
      "uuid": "668ca2ab-8d49-47f5-82a1-e3a82a58e5fb",
    },
  },
  LocalizedObject {
    "iso": "spa",
    "language": Object {
      "iso": "spa",
      "name": "Spanish",
    },
    "languages": Object {
      "eng": "English",
      "fre": "French",
      "spa": "Spanish",
    },
    "localized": true,
    "localizedKeys": Object {
      "name": true,
    },
    "localizedObject": Object {
      "destination_node_uuid": "bc978e00-2f3d-41f2-87c1-26b3f14e5925",
      "name": Array [
        "amarillo",
      ],
      "uuid": "14806949-d583-49e2-aa55-03aa16ee5a3a",
    },
  },
  LocalizedObject {
    "iso": "spa",
    "language": Object {
      "iso": "spa",
      "name": "Spanish",
    },
    "languages": Object {
      "eng": "English",
      "fre": "French",
      "spa": "Spanish",
    },
    "localized": true,
    "localizedKeys": Object {
      "name": true,
    },
    "localizedObject": Object {
      "destination_node_uuid": "bc978e00-2f3d-41f2-87c1-26b3f14e5925",
      "name": Array [
        "verde",
      ],
      "uuid": "77394377-f6b8-4366-9bef-d468621258ef",
    },
  },
  LocalizedObject {
    "iso": "spa",
    "language": Object {
      "iso": "spa",
      "name": "Spanish",
    },
    "languages": Object {
      "eng": "English",
      "fre": "French",
      "spa": "Spanish",
    },
    "localized": true,
    "localizedKeys": Object {
      "name": true,
    },
    "localizedObject": Object {
      "destination_node_uuid": "bc978e00-2f3d-41f2-87c1-26b3f14e5925",
      "name": Array [
        "azul",
      ],
      "uuid": "92d429d8-c275-4306-9360-93f4b9c7acb1",
    },
  },
  LocalizedObject {
    "iso": "spa",
    "language": Object {
      "iso": "spa",
      "name": "Spanish",
    },
    "languages": Object {
      "eng": "English",
      "fre": "French",
      "spa": "Spanish",
    },
    "localized": false,
    "localizedKeys": Object {},
    "localizedObject": Object {
      "destination_node_uuid": "bc978e00-2f3d-41f2-87c1-26b3f14e5925",
      "name": "Indigo",
      "uuid": "2de9af80-1bd9-4f37-839f-073edbd14369",
    },
  },
  LocalizedObject {
    "iso": "spa",
    "language": Object {
      "iso": "spa",
      "name": "Spanish",
    },
    "languages": Object {
      "eng": "English",
      "fre": "French",
      "spa": "Spanish",
    },
    "localized": false,
    "localizedKeys": Object {},
    "localizedObject": Object {
      "destination_node_uuid": "bc978e00-2f3d-41f2-87c1-26b3f14e5925",
      "name": "Violet",
      "uuid": "5760ec2f-04d4-492b-817b-9f395633ec79",
    },
  },
  LocalizedObject {
    "iso": "spa",
    "language": Object {
      "iso": "spa",
      "name": "Spanish",
    },
    "languages": Object {
      "eng": "English",
      "fre": "French",
      "spa": "Spanish",
    },
    "localized": false,
    "localizedKeys": Object {},
    "localizedObject": Object {
      "destination_node_uuid": "4fac7935-d13b-4b36-bf15-98075dca822a",
      "name": "Other",
      "uuid": "326a41b7-9bce-453b-8783-1113f649663c",
    },
  },
]
`;
